deepmd/pt/utils/utils.py

```diff
@@ -11,17 +11,162 @@
+) -> torch.Tensor:
+    # Tanh branch
+    tanh_term = torch.tanh(slope * (x - threshold))
+    grad_grad = -2 * slope * slope * tanh_term * (1 - tanh_term * tanh_term)
+
+    # SiLU branch
+    sig = 1.0 / (1.0 + torch.exp(-x))
+    sig_prime = sig * (1 - sig)
+    silu_term = sig_prime * (2 + x * (1 - 2 * sig))
+
+    grad_grad = torch.where(x >= threshold, grad_grad, silu_term)
+
+    return grad_output * grad_grad * grad_grad_output
+
+
+class CustomSiluFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x, threshold, slope, const_val):
+        ctx.save_for_backward(x)
+        ctx.threshold = threshold
+        ctx.slope = slope
+        ctx.const_val = const_val
+        return custom_silu_forward(x, threshold, slope, const_val)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        (x,) = ctx.saved_tensors
+        threshold = ctx.threshold
+        slope = ctx.slope
+
+        grad_input = CustomSiluGradFunction.apply(x, grad_output, threshold, slope)
+        return grad_input, None, None, None
+
+
+class CustomSiluGradFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x, grad_output, threshold, slope):
+        ctx.threshold = threshold
+        ctx.slope = slope
+        grad_input, grad = custom_silu_backward(x, grad_output, threshold, slope)
+        ctx.save_for_backward(x, grad_output, grad)
+        return grad_input
+
+    @staticmethod
+    def backward(ctx, grad_grad_output):
+        (x, grad_output, grad) = ctx.saved_tensors
+        threshold = ctx.threshold
+        slope = ctx.slope
+
+        grad_input = custom_silu_double_backward(
+            x, grad_grad_output, grad_output, threshold, slope
+        )
+        return grad_input, grad * grad_grad_output, None, None
+
+
+class CustomSiluScript(torch.nn.Module):
+    def __init__(self, threshold: float = 3.0):
+        super().__init__()
+        self.threshold = threshold
+
+        # Precompute parameters for the tanh replacement
+        sigmoid_threshold = 1 / (1 + np.exp(-threshold))
+        self.slope = float(
+            sigmoid_threshold + threshold * sigmoid_threshold * (1 - sigmoid_threshold)
+        )
+        self.const_val = float(threshold * sigmoid_threshold)
+
+    def forward(self, x):
+        return CustomSiluFunction.apply(x, self.threshold, self.slope, self.const_val)
+
+
+class CustomSilu(torch.nn.Module):
+    def __init__(self, threshold=3.0):
+        super().__init__()
+
+        def sigmoid(x):
+            return 1 / (1 + np.exp(-x))
+
+        def silu(x):
+            return x * sigmoid(x)
+
+        def silu_grad(x):
+            sig = sigmoid(x)
+            return sig + x * sig * (1 - sig)
+
+        self.threshold = threshold
+        self.slope = float(silu_grad(threshold))
+        self.const = float(silu(threshold))
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        silu_part = F.silu(x)
+        mask = x >= self.threshold
+        if torch.any(mask):
```

Comment:
Are you sure it is faster after adding the if-statement? This needs a device-to-host communication, so will block the CPU.

Reply:
When training 9 layers DPA3 on example water systems with single GPU, using if-statement takes **23s** per 100 steps versus **29s** without. It seems that this device-to-host communication is much cheaper than additional calculation.